pkg/ruler/manager.go

```diff
@@ -141,6 +141,9 @@ func (r *DefaultMultiTenantManager) syncRulesToManager(ctx context.Context, user
 		return
 	}
 
+	r.userManagerMtx.Lock()
+	defer r.userManagerMtx.Unlock()
```

Comment:
Moving it here improves but not fix the issue. We're still keeping the lock while calling `manager.Update()`, which could be very slow as described in #2280. Given the manager already has an internal mutex, we should just use `r.userManagerMtx` to protect the access to `r.userManagers` and to make sure we can't have race conditions creating 2 managers for the same tenant at the same time.

An idea could be refactoring a bit the code below. First we get the manager and create + start it if doesn't exist. This is done with the mutex taken. Then we update the manager (without taking the mutex).